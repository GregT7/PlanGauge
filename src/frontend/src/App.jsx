import TaskContextProvider from "../src/contexts/TaskContext";
import { ThemeProvider } from './components/ui/ThemeProvider'
import TaskTable from "./components/TaskTable"
import './App.css'
// import TestTable2 from "./components/TestTable2"

function App() {
  return (
      <ThemeProvider defaultTheme="dark" storageKey="vite-ui-theme" >
          <h1 className="pb-5 m-0 text-4xl leading-none">PlanGauge</h1>
            <TaskContextProvider>
              <TaskTable/>
            </TaskContextProvider>
        </ThemeProvider>
  );
}

export default App
