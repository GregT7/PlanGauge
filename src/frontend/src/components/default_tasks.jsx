export default [
  {
    id: 1,
    name: 'Plan Tool p58',
    category: 'Career',
    due_date: Date(2025, 6, 3),
    start_date: Date(2025, 6, 1),
    time_estimation: 90,
    selected: false
  },
  {
    id: 2,
    name: 'Update CV p6',
    category: 'Career',
    due_date: Date(2025, 6, 25),
    start_date: Date(2025, 6, 22),
    time_estimation: 45,
    selected: false
  },
  {
    id: 3,
    name: 'Plan Japan trip p2',
    category: 'Life',
    due_date: Date(2025, 7, 6),
    start_date: Date(2025, 6, 30),
    time_estimation: 90,
    selected: false
  },
];