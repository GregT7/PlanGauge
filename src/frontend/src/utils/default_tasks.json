[
  {
    "id": 1,
    "name": "Plan Tool p58",
    "category": "Career",
    "due_date": "2025-07-03",
    "start_date": "2025-07-01",
    "time_estimation": 90,
    "selected": false
  },
  {
    "id": 2,
    "name": "Update CV p6",
    "category": "Career",
    "due_date": "2025-07-25",
    "start_date": "2025-07-22",
    "time_estimation": 45,
    "selected": false
  },
  {
    "id": 3,
    "name": "Plan Japan trip p2",
    "category": "Life",
    "due_date": "2025-08-06",
    "start_date": "2025-07-30",
    "time_estimation": 90,
    "selected": false
  }
]
