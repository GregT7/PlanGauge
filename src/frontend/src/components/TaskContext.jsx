import { createContext, useState } from 'react';
import default_tasks from "./default_tasks";

export const TaskContext = createContext(undefined);


export default function TaskContextProvider({children}) {
    const [tasks, setTasks] = useState(default_tasks);

    return (
        <TaskContext.Provider value={{tasks, setTasks}}>
            {children}
        </TaskContext.Provider>
    );
}