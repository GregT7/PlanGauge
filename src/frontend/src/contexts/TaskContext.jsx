import { createContext, useState } from 'react';
import default_tasks from "../../public/default_tasks";

export const TaskContext = createContext(undefined);


export default function TaskContextProvider({children}) {
    const [tasks, setTasks] = useState(default_tasks);

    return (
        <TaskContext.Provider value={{tasks, setTasks}}>
            {children}
        </TaskContext.Provider>
    );
}