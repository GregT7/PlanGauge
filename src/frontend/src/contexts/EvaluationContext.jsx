import { useMemo, createContext, useContext } from "react";
import { TaskContext } from "./TaskContext";
import evaluatePlan from "@/utils/evaluatePlan";

export const EvaluationContext = createContext(undefined);

export default function EvaluationContextProvider({ children, statuses }) {
    const { tasks, timeSum } = useContext(TaskContext);
    const evaluation = useMemo(() => evaluatePlan(statuses, timeSum), 
        [tasks]
    );

    return (
        <EvaluationContext.Provider value={evaluation}>
            {children}
        </EvaluationContext.Provider>
    );
};