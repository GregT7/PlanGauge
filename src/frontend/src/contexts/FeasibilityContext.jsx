import { useMemo, useState, useEffect, createContext, useContext } from "react"
import defaultThresholds from "@/utils/defaultThresholds.json" with { type: 'json'}
import { TaskContext } from "./TaskContext"
import { StatsContext } from "./StatsContext"
import testCardData from "@/utils/testCardData.json" with { type: 'json'}
import toLocalMidnight from "@/utils/toLocalMidnight"
import { evaluateFeasibility } from "@/utils/evaluateFeasibility"

const calcStatus = (ave, std, sum) => {
  if (std === 0) return "undetermined";

  const zscore = Math.abs(sum - ave) / std;
  if (zscore <= 1) return "good";
  else if (zscore <= 2) return "moderate";
  else return "poor";
};

function isSameDay(d1, d2) {
  const day1 = toLocalMidnight(d1);
  const day2 = toLocalMidnight(d2);
  return day1.getTime() === day2.getTime();
}

function evalStatus(zScore, _std, sum) {
  zScore = Math.abs(zScore);
  if (sum === 0) return "neutral";
  else if (zScore > 2) return "poor";
  else if (zScore > 1) return "moderate";
  else if (zScore <= 1) return "good";
  else return "undefined";
}

export const feasibilityContext = createContext(undefined);

export function FeasibilityContextProvider({ children, cardData = testCardData }) {
const statsCtx = useContext(StatsContext);
const stats = statsCtx?.stats
const {tasks, timeSum} = useContext(TaskContext)

  const evaluatedCardSums = useMemo(() =>
    cardData.map(card => {
      const cardDate = toLocalMidnight(card.date);
      const daySum = tasks.reduce((sum, task) => {
        const taskDate = toLocalMidnight(task["start_date"]);
        return isSameDay(cardDate, taskDate) ? sum + task["time_estimation"] : sum;
      }, 0);
      return { ...card, sum: daySum };
    }),
    [tasks, cardData]
  );

  const evaluatedCardStatus = useMemo(() =>
    evaluatedCardSums.map(evalCard => {
      const zScore = (evalCard.sum - evalCard.ave) / evalCard.std;
      const newStatus = evalStatus(zScore, evalCard.std, evalCard.sum);
      return { ...evalCard, z_score: zScore, status: newStatus };
    }),
    [evaluatedCardSums]
  );

  const statusCount = useMemo(() => {
    const regex = /^(good|moderate|poor)$/;

    return evaluatedCardStatus.reduce((statObj, task) => {
      const statusValue = task?.status;
      if (statusValue && regex.test(statusValue)) {
        return {
          ...statObj,
          [statusValue]: statObj[statusValue] + 1,
        };
      }
      return statObj;
    }, { good: 0, moderate: 0, poor: 0 });
  }, [evaluatedCardStatus]);

const feasibility = useMemo(() => {
  const result = evaluateFeasibility(timeSum, stats?.week, statusCount, defaultThresholds);
  console.log("eval result:", result);
  return result;
}, [timeSum, statusCount, stats]);

  

//   console.log("evaluatedCardSums", evaluatedCardSums)
  console.log("evaluatedCardSums", statusCount)


    const test = 1
    return (
        <feasibilityContext.Provider value={test}>
            {children}
        </feasibilityContext.Provider>
    );     
}