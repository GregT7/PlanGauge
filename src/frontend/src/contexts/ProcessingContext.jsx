import { useMemo, useState, useEffect, createContext, useContext } from "react"
import defaultThresholds from "@/utils/defaultThresholds.json" with { type: 'json'}
import { TaskContext } from "./TaskContext"
import default_stats from '@/utils/default_stats';
import toLocalMidnight from "@/utils/toLocalMidnight"
import isSameDay from "@/utils/isSameDay";
import { evaluateFeasibility } from "@/utils/evaluateFeasibility"
import retrieveStats from '@/utils/retrieveStats';
import { genDefaultCardsData } from "@/utils/genDefaultCardData";
import { eval_status } from "@/utils/evaluateFeasibility";
import { toast } from 'sonner';
import updateCardStats from "@/utils/updateCardStats";

export const processingContext = createContext(undefined);

export function ProcessingContextProvider({children, starting_stats = default_stats, thresholds = defaultThresholds}) {
    const {tasks, timeSum} = useContext(TaskContext)

    // Generate starting card data for each day of the week
    let cardData = genDefaultCardsData()

    // Retrieve stats data
    const [stats, setStats] = useState(starting_stats);
    // const [cardData, setCardData] = useState(genDefaultCardsData())

    useEffect(() => {
        let cancelled = false;
            (async () => {
                try {
                const promise = retrieveStats(); // resolves to { message, details, data }

                // bind toast to the same promise (donâ€™t await the toast)
                toast.promise(promise, {
                    loading: 'Retrieving statistical data...',
                    success: (r) => r?.message ?? 'Stats loaded',
                    error:   (err) => err?.message || 'Failed to retrieve stats',
                });

                const res = await promise; // get full object back
                if (!cancelled && res?.data) setStats(res.data);
                } catch (err) {
                console.error('Stats init error:', err);
                if (!cancelled) toast.error('Error: Unexpected Internal Error When Retrieving Stats');
                }
            })();
    return () => { cancelled = true; };
    }, []);

    // Update Card Data
    const statsUpdated = updateCardStats(cardData, stats?.day)

    

    // On stat retrieval/update or task table update
    // 1. Process Card Data
    //      A. Calculate day sums for each card based on task state
    cardData = useMemo(() =>
        cardData.map(card => {
            const cardDate = toLocalMidnight(card.date);
            const daySum = tasks.reduce((sum, task) => {
                const taskDate = toLocalMidnight(task["start_date"]);
                return isSameDay(cardDate, taskDate) ? sum + task["time_estimation"] : sum;
            }, 0);
            return { ...card, sum: daySum };
        }),
        [tasks, cardData]
    );

    //      B. Calculate card status for each card based on task state
    // eval_category(sum, stats, zThresholds)
    cardData = useMemo(() =>
        cardData.map(evalCard => {
            const newStatus = eval_status(evalCard.sum, {std: evalCard.std, ave: evalCard.ave}, thresholds.zscore)
            return { ...evalCard, status: newStatus };
        }),
        [cardData]
    );


    //      C. Count the number of statuses
    const statusCount = useMemo(() => {
        const regex = /^(good|moderate|poor|unknown)$/;

        return cardData.reduce((statObj, task) => {
            const statusValue = task?.status;
            if (statusValue && regex.test(statusValue)) {
            return {
                ...statObj,
                [statusValue]: statObj[statusValue] + 1,
            };
            }
            return statObj;
        }, { good: 0, moderate: 0, poor: 0, unknown: 0});
    }, [cardData]);
    

    // 2. Process Overall Feasibility
    //      A. Calculate overall feasibility 
    //          Note: requires week_sum, week_stats, statusCount, thresholds
    const feasibility = useMemo(() => {
        return evaluateFeasibility(timeSum, stats?.week, statusCount, thresholds)
    }, [cardData, statusCount])

    //   const value = useMemo(() => ({ stats, setStats }), [stats]);
    const value = useMemo(() => ({stats, cardData, statusCount, feasibility, thresholds}), [stats, cardData])

    return (
        <processingContext.Provider value={value}>
            {children}
        </processingContext.Provider>
    );
}