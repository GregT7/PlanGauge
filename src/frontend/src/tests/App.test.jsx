import { describe, it, expect } from 'vitest';
// import userEvent from '@testing-lbrary/user-event';

import { screen, render } from '@testing-library/react';
import { useState } from 'react';
import { TaskContext } from '@/contexts/TaskContext';
import { vi } from 'vitest'
import default_tasks from '@/utils/default_tasks';
import App from "../App";

const AppWithContextWrapper = ({children}) => {
  const [tasks, setTasks] = useState(default_tasks);

  return (
    <TaskContext.Provider value={{tasks, setTasks}}>
      {children}
    </TaskContext.Provider>
  )
}

describe("App", () => {
  it("renders all components without problems", () => {

  });

  it("calls connectionTest only once at launch", () => {

  });

  it("displays Toaster component when toast is invoked", () => {

  });

  it("doesn't crash when connectionTest fails", () => {

  })
})

// describe("App integration tests", () => {

// })