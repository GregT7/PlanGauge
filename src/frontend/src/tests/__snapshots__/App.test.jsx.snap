// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App unit tests > passes snapshot test 1`] = `
<div>
  <h1
    class="pb-5 m-0 text-4xl leading-none"
  >
    PlanGauge
  </h1>
  <div
    class="relative w-full overflow-x-auto"
    data-slot="table-container"
  >
    <table
      class="caption-bottom text-sm w-[73.5%] mx-auto"
      data-slot="table"
    >
      <thead
        class=""
        data-slot="table-header"
      >
        <tr
          class="data-[state=selected]:bg-muted transition-colors"
          data-slot="table-row"
        >
          <th
            class="text-foreground h-10 px-2 text-left align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] w-[3.5%]"
            data-slot="table-head"
          />
          <th
            class="text-foreground h-10 align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-left px-2 border border-stone-600 w-[30%]"
            data-slot="table-head"
          >
            Task Name
          </th>
          <th
            class="text-foreground h-10 align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-left px-2 border border-stone-600 w-[12.5%]"
            data-slot="table-head"
          >
            Category
          </th>
          <th
            class="text-foreground h-10 align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-left px-2 border border-stone-600 w-[12.5%]"
            data-slot="table-head"
          >
            Due Date
          </th>
          <th
            class="text-foreground h-10 align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-left px-2 border border-stone-600 w-[12.5%]"
            data-slot="table-head"
          >
            Start Date
          </th>
          <th
            class="text-foreground h-10 align-middle font-medium whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-left px-2 border border-stone-600 w-[2.5%]"
            data-slot="table-head"
          >
            Time
          </th>
        </tr>
      </thead>
      <tbody
        class=""
        data-slot="table-body"
      >
        <tr
          class="data-[state=selected]:bg-muted transition-colors"
          data-slot="table-row"
        >
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-role="row-selector"
            data-slot="table-cell"
          >
            <button
              aria-checked="false"
              class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 cursor-pointer"
              data-slot="checkbox"
              data-state="unchecked"
              role="checkbox"
              type="button"
              value="on"
            />
          </td>
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <input
              class="outline-none w-full h-full bg-transparent text-left px-2"
              type="text"
              value="Plan Tool p58"
            />
          </td>
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] px-2 border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-expanded="false"
              aria-haspopup="menu"
              class="cursor-pointer w-full h-full text-left"
              data-slot="dropdown-menu-trigger"
              data-state="closed"
              id="radix-:r0:"
              type="button"
            >
              <span
                class="cursor-pointer bg-amber-600 px-1.5 py-0.5 rounded-sm"
              >
                Career
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-controls="radix-:r2:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="w-full h-full text-left cursor-pointer"
              data-slot="popover-trigger"
              data-state="closed"
              type="button"
            >
               
              <span
                class="px-2 py-2 wrap-normal"
              >
                 
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-controls="radix-:r3:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="w-full h-full text-left cursor-pointer"
              data-slot="popover-trigger"
              data-state="closed"
              type="button"
            >
               
              <span
                class="px-2 py-2 wrap-normal"
              >
                 
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <input
              class="text-left px-2 w-full h-full bg-transparent outline-none [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
              type="number"
              value="90"
            />
          </td>
        </tr>
        <tr
          class="data-[state=selected]:bg-muted transition-colors"
          data-slot="table-row"
        >
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-role="row-selector"
            data-slot="table-cell"
          >
            <button
              aria-checked="false"
              class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 cursor-pointer"
              data-slot="checkbox"
              data-state="unchecked"
              role="checkbox"
              type="button"
              value="on"
            />
          </td>
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <input
              class="outline-none w-full h-full bg-transparent text-left px-2"
              type="text"
              value="Update CV p6"
            />
          </td>
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] px-2 border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-expanded="false"
              aria-haspopup="menu"
              class="cursor-pointer w-full h-full text-left"
              data-slot="dropdown-menu-trigger"
              data-state="closed"
              id="radix-:r4:"
              type="button"
            >
              <span
                class="cursor-pointer bg-amber-600 px-1.5 py-0.5 rounded-sm"
              >
                Career
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-controls="radix-:r6:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="w-full h-full text-left cursor-pointer"
              data-slot="popover-trigger"
              data-state="closed"
              type="button"
            >
               
              <span
                class="px-2 py-2 wrap-normal"
              >
                 
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-controls="radix-:r7:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="w-full h-full text-left cursor-pointer"
              data-slot="popover-trigger"
              data-state="closed"
              type="button"
            >
               
              <span
                class="px-2 py-2 wrap-normal"
              >
                 
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <input
              class="text-left px-2 w-full h-full bg-transparent outline-none [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
              type="number"
              value="45"
            />
          </td>
        </tr>
        <tr
          class="data-[state=selected]:bg-muted transition-colors"
          data-slot="table-row"
        >
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-role="row-selector"
            data-slot="table-cell"
          >
            <button
              aria-checked="false"
              class="peer border-input dark:bg-input/30 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground dark:data-[state=checked]:bg-primary data-[state=checked]:border-primary focus-visible:border-ring focus-visible:ring-ring/50 aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive size-4 shrink-0 rounded-[4px] border shadow-xs transition-shadow outline-none focus-visible:ring-[3px] disabled:cursor-not-allowed disabled:opacity-50 cursor-pointer"
              data-slot="checkbox"
              data-state="unchecked"
              role="checkbox"
              type="button"
              value="on"
            />
          </td>
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <input
              class="outline-none w-full h-full bg-transparent text-left px-2"
              type="text"
              value="Plan Japan trip p2"
            />
          </td>
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] px-2 border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-expanded="false"
              aria-haspopup="menu"
              class="cursor-pointer w-full h-full text-left"
              data-slot="dropdown-menu-trigger"
              data-state="closed"
              id="radix-:r8:"
              type="button"
            >
              <span
                class="cursor-pointer bg-teal-600 px-1.5 py-0.5 rounded-sm"
              >
                Life
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-controls="radix-:ra:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="w-full h-full text-left cursor-pointer"
              data-slot="popover-trigger"
              data-state="closed"
              type="button"
            >
               
              <span
                class="px-2 py-2 wrap-normal"
              >
                 
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <button
              aria-controls="radix-:rb:"
              aria-expanded="false"
              aria-haspopup="dialog"
              class="w-full h-full text-left cursor-pointer"
              data-slot="popover-trigger"
              data-state="closed"
              type="button"
            >
               
              <span
                class="px-2 py-2 wrap-normal"
              >
                 
              </span>
            </button>
          </td>
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 text-left border border-stone-600 hover:bg-stone-800"
            data-slot="table-cell"
          >
            <input
              class="text-left px-2 w-full h-full bg-transparent outline-none [appearance:textfield] [&::-webkit-outer-spin-button]:appearance-none [&::-webkit-inner-spin-button]:appearance-none"
              type="number"
              value="90"
            />
          </td>
        </tr>
      </tbody>
      <tfoot
        class="font-medium"
        data-slot="table-footer"
      >
        <tr
          class="data-[state=selected]:bg-muted transition-colors"
          data-slot="table-row"
        >
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            data-slot="table-cell"
          />
          <td
            class="align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] h-10 border border-stone-600"
            colspan="5"
            data-slot="table-cell"
          >
            <button
              class="text-left hover:bg-stone-800 w-full h-full px-2"
              data-testid="add-task-button"
            >
              + New Page
            </button>
          </td>
        </tr>
        <tr
          class="data-[state=selected]:bg-muted transition-colors"
          data-slot="table-row"
        >
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px]"
            colspan="5"
            data-slot="table-cell"
          />
          <td
            class="h-10 align-middle whitespace-nowrap [&:has([role=checkbox])]:pr-0 [&>[role=checkbox]]:translate-y-[2px] text-right"
            data-slot="table-cell"
          >
            <span
              class="text-stone-400 px-2"
            >
              SUM
            </span>
            <span
              data-testid="time-display"
            >
              225
            </span>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
